### ISSUE TITLE
urlAuthentication with redirect on IndexRoute

### ISSUE THREAD

**Issue by kevinansfield** [heartsentwined/ember-auth#37]
I'm trying to implement urlAuthentication so that I can authenticate users immediately after hitting the confirmation link generated by devise but I'm running into an issue.

When I hit the url `/?auth[key]=xxx/signup/step_2`, instead of auth happening and being taken to the `/signup/step_2` url I instead get immediately redirected by the `redirect` set on `App.IndexRoute`.

**Comment by kevinansfield** [comment]
> URL authentication will not take effect if the entry point itself is not an Auth.Route: the user will not be signed in; (hence) no events will fire.

Does that mean that the `App.IndexRoute` needs to be an `Auth.Route`?

**Comment by kevinansfield** [comment]
Don't worry, I think I've figured this out now.

I had to add an `AuthRoute` that extends `Auth.Route` and change my confirmation redirect to `/auth?auth[key]=xxx/signup/step_2`

It may be useful for other users if it's made a little clearer in the docs that something like this is necessary if you have `authRedirect: true` and you don't want your app's homepage to require authentication.

**Comment by kevinansfield** [comment]
Well, I thought I was there but it appears that the urlAuthentication isn't passing the params correctly. Here's an excerpt from the development.log:

```
Started GET "/auth?auth[key]=2DmAKBy3vihL3mBm3MxX/signup/step_2" for 127.0.0.1 at 2013-04-17 17:30:15 +0200
Processing by StaticController#app as HTML
  Parameters: {"auth"=>{"key"=>"2DmAKBy3vihL3mBm3MxX/signup/step_2"}, "ember"=>"auth"}
  User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" = 15 LIMIT 1
  Rendered static/app.html.erb within layouts/application (0.1ms)
Completed 200 OK in 1213ms (Views: 1211.3ms | ActiveRecord: 0.3ms)


Started POST "/users/sign_in" for 127.0.0.1 at 2013-04-17 17:30:18 +0200
Processing by SessionsController#create as JSON
  Parameters: {"auth"=>{"undefined"=>"undefined"}, "session"=>{"auth"=>{"undefined"=>"undefined"}}}
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = 15 LIMIT 1
Completed 400 Bad Request in 3ms (Views: 0.2ms | ActiveRecord: 0.2ms)
```

**Comment by heartsentwined** [comment]
Two separate issues. First, I managed to reproduce the `undefined` object params bug, and I'm working on it. As for the site-wide url authentication, this is also being implemented, as an extension of #36. The result will be a flag that you can turn on, in order to enable url authenticate on regular `Em.Route` too.

**Comment by kevinansfield** [comment]
Awesome. Thank you!

**Comment by heartsentwined** [comment]
new version released. It fixes the `undefined` bug. As for enabling site-wide Url Authentication, set this flag:

(section from [readme](https://github.com/heartsentwined/ember-auth#url-authentication)):

> By default, url authentication is only available on `Auth.Route`s. If you
> want it to be available for regular `Em.Route`s too, set:
> 
> ``` coffeescript
> Auth.Config.reopen
>   urlAuthenticationRouteScope: 'both' # defaults to 'auth'
> ```

This will enable your non-`Auth.Route` Index Route to work with url authentication.

The original idea to put this functionality only in `Auth.Route` is the reasoning that public routes don't need authenticating, and that url authentication is only intended to enable Direct Linking to some protected areas. But your use case is solid too. We'll see if your use case appears to be the majority - in that case we could reverse the default setting of `ember-auth` to make things easier for newer devs.

**Comment by kevinansfield** [comment]
I've just updated and added the urlAuthenticationRouteScope config but now there is a different issue.

If I load the app on the /home route I see two failed sign in attempts with the parameters:

`Parameters: {"auth"=>{"undefined"=>"undefined"}, "session"=>{"auth"=>{"undefined"=>"undefined"}}}`

After this, every time a new route is loaded there is an additional sign in attempt with `undefined` in the parameters.

**Comment by heartsentwined** [comment]
Sorry about that, are you using `4.1.0` or `4.1.1`? The `4.1.0` version had a bug during build; I have replaced it with `4.1.1` immediately.

Reopening this issue until all your issues have been resolved, and that "everything is working".

**Comment by kevinansfield** [comment]
I've just updated to `4.1.1` but I'm still seeing the same issue. Is there any other details I can provide that will help track the problem down?

**Comment by heartsentwined** [comment]
Your `Auth.Config` settings, whether you are using `history` or `hash` routing (if you don't know, then default is `hash`), and the exact full URLs that you are trying to authenticate on, please.

**Comment by heartsentwined** [comment]
One thing I noticed: You said you were trying to sign in to `/?auth[key]=xxx/signup/step_2`. This is neither a proper `history` route nor a `hash` location route. The example given on the doc uses the `hash` route, and in your case would take the form of `/?auth[key]=xxx/#/signup/step_2`; as for a history route, it would take the form of `/signup/step_2?auth[key]=xxx`. Both would pass a param of `auth[key]=xxx` (trailing slash stripped from `hash` case), and ember would pass both URL as `/signup/step_2`.

As for what `/?auth[key]=xxx/signup/step_2` means, I think this is just invalid URL. Either you mean a full path with `?auth[key]xxx` as one of the paths (it should be percent-quoted then), or you mean `auth[key]` to have a value of `xxx/signup/step_2` (it should still be percent-quoted).

Not sure if this is the cause, just what I have spotted based on what information we have here.

**Comment by kevinansfield** [comment]
```
Auth.Config.reopen
  tokenCreateUrl: '/users/sign_in'
  tokenDestroyUrl: '/users/sign_out'
  tokenKey: 'auth_token'

  idKey: 'user_id'
  userModel: App.User

  rememberMe: true
  rememberTokenKey: 'remember_token'
  rememberPeriod: 14 # days
  # rememberAutoRecallRouteScope: 'both'

  authRedirect: true
  smartSignInRedirect: true
  signInRoute: 'sign_in'
  signOutRoute: 'sign_out'
  signInRedirectFallbackRoute: 'newsfeed'
  signOutRedirectFallbackRoute: 'home'

  urlAuthentication: true
  urlAuthenticationParamsKey: 'auth'
  urlAuthenticationRouteScope: 'both'
```

I'm using `location` for routing. You're correct, I just took the `hash` example and mangled it when I tried to convert it to the `location` version.

However, the problem now is not when trying to authenticate, it's when viewing any route.

**Comment by heartsentwined** [comment]
`location`? You mean `history`?

I'll try to hack together an example app to reproduce your environment. If you have a repo of this example app, would you mind sharing it? Otherwise, I'll try to describe what I have gathered from what you have said, and you can verify if I got it right.

**Comment by kevinansfield** [comment]
> `location`? You mean `history`?

Yes, sorry.

Unfortunately I'm unable to share the repository. I can see if I can set up a minimal app that shows the same issue if that's necessary.

**Comment by kevinansfield** [comment]
This may be interesting. Despite opening just a basic route `/home`, the `params` object is not empty:

![Screen Shot 2013-04-17 at 19 18 32](https://f.cloud.github.com/assets/415/392627/e1218b90-a782-11e2-8208-2a5c2a02d0b0.png)

**Comment by heartsentwined** [comment]
`Object { "undefined": "undefined" }` bug fixed in `4.1.2`. Many thanks for this bug report!

See if this helps your problem. I know it may just be a peripheral issue, but I can't reproduce the behavior you have described yet.

**Comment by kevinansfield** [comment]
That's fixed the issue, thanks!
